### ISSUE TITLE
Dose faiss support user defined distance function?

### ISSUE THREAD

**Issue by xieyi318** [facebookresearch/faiss#638]
Dose faiss support customized distance function as (metric='pyfunc') do in sklearn.neighbors.NearestNeighbors() ?
Thanks a lot!

**Comment by beauby** [comment]
No, the only "distance" functions supported are L2 and inner product.

**Comment by jpetot** [comment]
Same question, is it still not possible to use custom distance function ? 
Many thanks if it would be possible !

**Comment by phoenixzw** [comment]
Hi @jpetot , did you figured out a solution to use user defined distance function? I meet the same problem.
Thanks!

**Comment by Enet4** [comment]
This seems like a fairly frequent question, but it makes sense that this functionality is not present. Most dense vector indexing algorithms only make sense under a specific distance function. Even if Faiss enabled users to set an arbitrary distance function, the various index types and index splitting solutions available (HNSW, LSH, IVF, ...) rely on assumptions that would likely be broken by the distance function, thus returning poor search results.

At best, custom distance functions could only be supported by `IndexFlat`, but there is a cost in this form of integration, especially if they need to cross the environment boundary, as would be the case in Python. This frequent operation is not something that can be done with zero overhead for Python functions, so performance would also be poor.

When having to employ a custom metric, there are two options:

- Transform the vector space so that one of the supported metrics becomes equivalent or sufficiently similar to the distance function in the original vector space.
- Not use Faiss: instead keep all vectors in a contiguous ndarray and perform exhaustive k-NN manually.

**Comment by dbickson** [comment]
Today I asked openAI chatGPT on how to define custom distance function at faiss and the chat bot invented the following code: 
```


#include <faiss/IndexFlat.h>
#include <algorithm>
#include <cmath>

// Define a custom distance metric function that takes two vectors as input
// and returns the distance between them
float custom_distance(int d, const float *x, const float *y) {
    float sum = 0;
    for (int i = 0; i < d; i++) {
        sum += std::abs(x[i] - y[i]);
    }
    return sum;
}

int main() {
    // Create a Faiss index for the nearest neighbor model
    faiss::IndexFlat index(128);

    // Set the custom distance metric function for the index
    index.set_distance(custom_distance);

    // Train the nearest neighbor model using the custom distance metric
    index.train(data);

    return 0;
}
```
I think it is a very clean interface, users who asked for custom distance are not concerned about performance but are concorned of usability it would be great to support something similar even if the implementation is not a blas/ block implementation.

**Comment by Enet4** [comment]
@dbickson ChatGPT is not a reliable source of code solutions, nor does it serve well as an unbiased and reasonable decision maker in terms of library and API design.

With that said, you could use whichever automated tools are at your disposal to help implement a custom vector search on nparrays. Faiss is still not quite fit for the use case.
